website/server/models/user/methods.js

```diff
@@ -20,6 +20,49 @@ schema.methods.getGroups = function getUserGroups () {
   return userGroups;
 };
 
+// Get an array of error message keys that would be thrown if the given interaction was attempted
+schema.methods.getObjectionsToInteraction = function getObjectionsToInteraction (interaction, receiver) {
+  let sender = this;
+
+  /* eslint-disable no-unused-vars */ // The checks below all get access to sndr and rcvr, but not all use both
+  let checks = {
+    always: [
+      // Revoked chat privileges block all interactions to prevent the evading of harassment protections
+      // See issue #7971 for some discussion
+      (sndr, rcvr) => sndr.flags.chatRevoked && 'chatPrivilegesRevoked',
+
+      // Direct user blocks prevent all interactions
+      (sndr, rcvr) => rcvr.inbox.blocks.includes(sndr._id) && 'notAuthorizedToSendMessageToThisUser',
+      (sndr, rcvr) => sndr.inbox.blocks.includes(rcvr._id) && 'notAuthorizedToSendMessageToThisUser',
+    ],
+
+    'send-private-message': [
+      // Private messaging has an opt-out, which does not affect other interactions
+      (sndr, rcvr) => rcvr.inbox.optOut && 'notAuthorizedToSendMessageToThisUser',
+
+      // We allow a player to message themselves so they can test how PMs work or send their own notes to themselves
+    ],
+
+    'transfer-gems': [
+      // Unlike private messages, gems can't be sent to oneself
+      (sndr, rcvr) => rcvr._id === sndr._id && 'cannotSendGemsToYourself',
+    ],
+  };
+  /* eslint-enable no-unused-vars */
+
+  let knownInteractions = Object.keys(checks).filter((key) => key !== 'always');
+
+  if (!knownInteractions.includes(interaction)) {
+    throw new Error(`Unknown kind of interaction: "${interaction}", expected one of ${knownInteractions.join(', ')}`);
+  }
+
+  let checksToRun = checks.always.concat(checks[interaction]);
```

Comment:
I've converted it to lodash. I'm fairly new to JS and haven't used that library before, so I hope I did it the right way. :)

TBH, I'm not sure I'm that concerned about the arrays; the number is constant per call and I don't think the method is going to be a bottleneck, as it is likely only to be called in response to relatively infrequent UI events.

I realize I've gone against some of your requests, I've considered them over and over again and couldn't get myself to agree with them. Apologies if this causes any issues and thank you for your reviews. Also apologies for the huge delay... it took me being shocked from recent events to numb my fear of this conversation enough... :/ I'll gladly step back if you'd prefer someone else to work on this.